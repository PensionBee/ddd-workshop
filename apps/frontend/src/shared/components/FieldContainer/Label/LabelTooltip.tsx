import { LargeSystemIcon } from "@shared/components/Svg";
import { Tooltip } from "@shared/components/Tooltip";

export type LabelTooltipProps = {
  labelTooltip: string;
};

export const LabelTooltip: React.FC<LabelTooltipProps> = ({
  labelTooltip,
  ...props
}) => (
  <Tooltip content={labelTooltip} {...props}>
    <LargeSystemIcon icon="Info" fill="grey2" w={24} h={24} />
  </Tooltip>
);
