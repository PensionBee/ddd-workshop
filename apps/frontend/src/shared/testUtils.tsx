import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
