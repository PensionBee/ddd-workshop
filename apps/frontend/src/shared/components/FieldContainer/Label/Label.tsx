import { useMemo } from "react";
import { Box } from "@shared/components/Box";
import { Link } from "@shared/components/Link";
import { BodySmallB } from "@shared/components/Typography";
import { LabelTooltip } from "./LabelTooltip";

export type LabelProps = {
  /** Label for field */
  label?: string;
  /** Label 'i' tooltip */
  labelTooltip?: React.ComponentProps<typeof LabelTooltip>["labelTooltip"];
  /** Label link href */
  labelHref?: string;
  /** Label link text */
  labelLinkText?: string;
  /** Label htmlFor - id of input */
  labelHtmlFor?: string;
};

export const Label: React.FC<LabelProps> = ({
  label,
  labelTooltip,
  labelHref,
  labelLinkText,
  labelHtmlFor,
  ...props
}) => {
  const conditionalContent = useMemo(() => {
    if (labelTooltip) return <LabelTooltip labelTooltip={labelTooltip} />;
    if (labelHref) return <Link href={labelHref}>{labelLinkText}</Link>;
    return null;
  }, [labelTooltip, labelHref, labelLinkText]);

  if (!label) return null;

  return (
    <Box
      as="label"
      htmlFor={labelHtmlFor}
      d="flex"
      justify="space-between"
      align="start"
      pb={16}
      colGap={16}
      mt={0}
      nowrap
      {...props}
    >
      <BodySmallB>{label}</BodySmallB>
      {conditionalContent}
    </Box>
  );
};
