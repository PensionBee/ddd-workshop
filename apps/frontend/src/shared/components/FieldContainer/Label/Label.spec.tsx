import { render, vi } from "@shared/testUtils";
import { Label } from "./Label";

const LABEL_TEXT = "Label text";

describe("FieldWrapper/Label", () => {
  it("should render correctly", () => {
    const { getByText, asFragment } = render(<Label label={LABEL_TEXT} />);

    expect(getByText(LABEL_TEXT)).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });
  it("should not render if no label", () => {
    const { container } = render(<Label />);

    expect(container.firstChild).toBeNull();
  });
});
