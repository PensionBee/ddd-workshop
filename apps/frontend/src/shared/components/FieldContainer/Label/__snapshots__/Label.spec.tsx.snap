// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FieldWrapper/Label > should render correctly 1`] = `
<DocumentFragment>
  <label
    class="sc-bczSft jJehPQ"
  >
    <span
      class="sc-dkzCjb cGViLk"
      data-rc="BodySmallB"
    >
      Label text
    </span>
  </label>
</DocumentFragment>
`;

exports[`FieldWrapper/Label should render correctly 1`] = `
<DocumentFragment>
  <label
    class="sc-bcXHqe bjGbzT"
  >
    <span
      class="sc-dkrFOg lcwtgS"
      data-rc="BodySmallB"
    >
      Label text
    </span>
  </label>
</DocumentFragment>
`;
